horovod/mxnet/mpi_ops.cc

```diff
@@ -0,0 +1,335 @@
+      GetOpNameHandle("broadcast", name, handle), device,
+      [handle](const Status& status) {
+        handle_manager.MarkDone(handle, status);
+        handle_manager.ExecuteCallback(handle);
+      });
+  ThrowIfError(enqueue_result);
+
+  return handle;
+}
+
+#if HAVE_CUDA
+int DoBroadcastCudaOnCPU(
+    std::shared_ptr<MXTemporaryBuffer<NDArray>>& hvd_cpu_buffer, int root_rank,
+    std::string& name, Callback cb) {
+  // Make async copy of input tensor to CPU tensor and record completion event.
+  auto hvd_context = std::make_shared<MXOpContext<NDArray>>(
+      CPU_DEVICE_ID, hvd_cpu_buffer->tensor());
+  auto ready_event =
+      std::make_shared<MXReadyEvent<NDArray>>(hvd_cpu_buffer->tensor());
+
+  auto handle = handle_manager.AllocateHandle(cb);
+  auto enqueue_result = EnqueueTensorBroadcast(
+      hvd_context, hvd_cpu_buffer, hvd_cpu_buffer, root_rank, ready_event,
+      GetOpNameHandle("broadcast", name, handle), CPU_DEVICE_ID,
+      [handle](const Status& status) {
+        handle_manager.MarkDone(handle, status);
+        handle_manager.ExecuteCallback(handle);
+      });
+  ThrowIfError(enqueue_result);
+
+  return handle;
+}
+#endif
+
+extern "C" int horovod_mxnet_allreduce_async(NDArray* input, NDArray* output,
+                                             char* name, bool average) {
+
+  std::string new_name = GetOpName("allreduce", name);
+  auto allreduce_async_fn = [input, output,
+                             new_name](RunContext rctx,
+                                       Engine::CallbackOnComplete cb) mutable {
+    DoAllreduce(input, output, new_name, cb);
+  };
+#if HAVE_CUDA
+  auto allreduce_async_cpu_fn =
+      [input, output, new_name](RunContext rctx,
+                                Engine::CallbackOnComplete cb) mutable {
+        DoAllreduceCudaOnCPU(input, output, new_name, cb);
+      };
+#endif
+
+#if HAVE_CUDA && !HOROVOD_GPU_ALLREDUCE
+  // Not in-place
+  if (input->var() != output->var()) {
+    Engine::Get()->PushAsync(allreduce_async_cpu_fn, input->ctx(),
+                             {input->var()}, {output->var()},
+                             FnProperty::kNormal, 0, "HorovodAllreduce");
+    // In-place
+  } else {
+    Engine::Get()->PushAsync(allreduce_async_cpu_fn, input->ctx(), {},
+                             {output->var()}, FnProperty::kNormal, 0,
+                             "HorovodAllreduce");
+  }
+#else
+  // Not in-place
+  if (input->var() != output->var()) {
+    Engine::Get()->PushAsync(allreduce_async_fn, input->ctx(), {input->var()},
+                             {output->var()}, FnProperty::kNormal, 0,
+                             "HorovodAllreduce");
+    // In-place
+  } else {
+    Engine::Get()->PushAsync(allreduce_async_fn, input->ctx(), {},
+                             {output->var()}, FnProperty::kNormal, 0,
+                             "HorovodAllreduce");
+  }
+#endif
+
+  if (average) {
+    *output /= horovod_size();
```

Comment:
Is it truly async function in a sense that control would be returned before CUDA kernels are done?  If so, would `*output /= horovod_size()` would also be enqueued operation and return immediately?  How do we guarantee that it will execute after allreduce, since allreduce is using NCCL CUDA stream?

In PyTorch, we put synchronous averaging in callback that is passed to `EnqueueTensorAllreduce()`.

Reply:
Yes, it is truly async in that sense. The `/=` operator calls down to a `PushSync` operation that pushes it to the engine, and the allreduce calls a `PushAsync` operation. Despite the naming, both operations are asynchronous in the control returning sense you are talking about as described here [1]:

> All Push APIs are asynchronous. The API call returns immediately regardless of whether the pushed Fn is finished or not.

We can guarantee that it will execute after allreduce, because the callback will be called after the Horovod MPI call finishes and only then will the engine release the read lock on output.

[1] https://mxnet.incubator.apache.org/architecture/overview.html